stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,367 @@
+# Copyright 2015 Rackspace, Inc.
+# Copyright 2018 StackHPC Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import glob
+import itertools
+import os
+from collections import namedtuple
+
+from ironic_python_agent import errors
+from ironic_python_agent import hardware
+from ironic_python_agent import utils
+
+from oslo_concurrency import processutils
+from oslo_log import log
+from oslo_utils import strutils
+
+_MISSING_MATCHER_RULE = (
+    "Not checking NIC firmware for {interface}. "
+    "There was no rule for pci_id: {pci_id} set in "
+    "nic_firmware property"
+)
+_DUPLICATE_ENTRIES_MSG_TEMPLATE = (
+    "Multiple matchers with the same pci_id. The conflict "
+    "occured for pci_id: {pci_id}. Using "
+    "firmware_version: {new} instead of {old}"
+)
+_FIRMWARE_CHECK_DISABLE_KEY = 'disable_nic_firmware_check'
+_HELP_MSG_EXAMPLE = (
+    "For cleaning to proceed, you must set the property "
+    "'nic_firmware' in the node's extra field, for example:\n"
+    "$ openstack baremetal node set $NODE_ID --extra "
+    """nic_firmware='[{"vendor_id": "15B3","device_id": " "1013","""
+    """firmware_version": "12.20.1019"}]'"""
+)
+
+_UEVENT_PCI_SLOT_NAME_PREFIX = "PCI_SLOT_NAME="
+
+LOG = log.getLogger()
+
+
+def _get_base_in_relative_path(path):
+    # given relative path, returns the base element, e.g one/two/three,
+    # returns one
+    if os.path.isabs(path):
+        raise ValueError("path must be relative")
+    return path.split(os.sep)[0]
+
+
+def _parse_uevent(interface_line_pairs):
+    # converts the contents of /sys/class/net/*/device/uevent into a dictionary
+    # The name in the wild card position becomes the key and contents of uevent
+    # is itself mapped to a dictionary.
+    #
+    # the uevent dictionary maps the component before the first equals sign to
+    # to the component after e.g given PCI_SLOT_NAME=0000:00:19.0\n,
+    # the dictionary will contain: {"PCI_SLOT_NAME" : "0000:00:19.0"}
+    result = {}
+    for interface, _line in interface_line_pairs:
+        line = iter(_line.rstrip())
+        if interface not in result:
+            result[interface] = {}
+        split = itertools.takewhile(lambda x: x != "=", line)
+        key = "".join(split)
+        value = "".join(line)
+        result[interface][key] = value
```

Comment:
Won't we end up with the key in the value here? e.g. "PCI_SLOT_NAME=0000:00:19.0"

Reply:
No, since the iterator is partially exhausted by `takewhile`, it just contains the remainder. Maybe line wasn't the best name...